src/webextension/list/filter.js

```diff
@@ -7,29 +7,56 @@ export function parseFilterString(filter) {
   if (!filter) {
     return [];
   }
-  return filter.split(/\s+/).map((i) => i.toLocaleLowerCase());
+  return filter.split(/\s+/);
+}
+
+function tryMakeURL(string, defaultProtocol = null) {
+  let url = null;
+  try {
+    url = new URL(string);
+  } catch (e) {
+    if (defaultProtocol) {
+      try {
+        url = new URL(defaultProtocol + "//" + string);
+      } catch (e) {}
+    }
+  }
+  return {url, string};
 }
 
 function match(filterElement, value) {
-  return value.includes(filterElement);
+  return value.toLocaleLowerCase().includes(filterElement.toLocaleLowerCase());
+}
+
+function matchURL(filterElement, value) {
+  // If the user is in the middle of typing a URL or entered something not
+  // URL-like in the origin field, just to a string comparison. This is safe for
+  // strictly matching auto-filled URLs since they're always full hosts with
+  // protocol (and possibly port).
+  if (!filterElement.url || !value.url) {
+    return match(filterElement.string, value.string);
+  }
+  return (filterElement.url.protocol === value.url.protocol &&
+          filterElement.url.host === value.url.host);
 }
 
-function matchAny(filterElement, values) {
+function matchAny(filterElement, values, matcher = match) {
   for (let i of values) {
-    if (match(filterElement, i)) {
+    if (matcher(filterElement, i)) {
       return true;
     }
   }
   return false;
 }
 
 export function filterItem(filter, item) {
```

Comment:
Looking at these changes, my sense is filtering needs to get more differentiating -- instead of relying on a string that is (potentially mis)interpreted in a few ways, it's an object that better outlines what the filter intent is.

Reply:
I'm not sure what this means. What *part* of filtering needs to be converted from a string to an object? The visible representation of a filter should be a string, since it's what goes into the text field, and one of the main reasons for having filters in the first place is as something a user could type.

If you're suggesting that the automatically-generated filter should only find results in the `origin` field of the item, I agree and that's #627. I'd like it to be something like Google's mini-format for search queries, e.g. `url:https://www.example.com` (there'd be other prefixes too, like `title:`, `username:`, and maybe one day, `tag:`).